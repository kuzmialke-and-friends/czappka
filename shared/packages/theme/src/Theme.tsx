import React, { createContext, FunctionComponent, useContext, useEffect, useState } from 'react';

const themes = {
    dark: {
        color: '#A3B4C8',
        'highlight-color-1': '#CC6D3D',
        'highlight-color-2': '#B74803',
        'background-color': '#022E51',
    },
    light: {
        color: '#285185',
        'highlight-color-1': '#D67940',
        'highlight-color-2': '#6F4849',
        'background-color': '#CCD9E2',
    },
};

type Theme = keyof typeof themes;
type Variables = typeof themes.dark & typeof themes.light;

const ThemeSelectorContext = createContext({
    themeName: 'dark',
    toggleTheme: () => {},
});

export const useTheme = () => useContext(ThemeSelectorContext);

const setCSSVariables = (theme: Variables) => {
    Object.entries(theme).forEach(([variableName, variable]) => {
        document.documentElement.style.setProperty(`--${variableName}`, variable);
    });
};

export const ThemeProvider: FunctionComponent = ({ children }) => {
    const [themeName, setThemeName] = useState<Theme>('dark');
    const [theme, setTheme] = useState(themes[themeName]);

    useEffect(() => {
        setCSSVariables(theme);
    }, [theme]);

    const toggleTheme = () => {
        if (theme === themes.dark) {
            setTheme(themes.light);
            setThemeName('light');
        } else {
            setTheme(themes.dark);
            setThemeName('dark');
        }
    };

    return <ThemeSelectorContext.Provider value={{ toggleTheme, themeName }}>{children}</ThemeSelectorContext.Provider>;
};
