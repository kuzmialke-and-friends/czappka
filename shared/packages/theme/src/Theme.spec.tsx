import React from 'react';
import renderer from 'react-test-renderer';
import { renderHook, cleanup } from '@testing-library/react-hooks';
import { useTheme, ThemeProvider } from './Theme';

afterEach(cleanup);

describe('theme', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(
                <ThemeProvider>
                    <div>test</div>
                </ThemeProvider>,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('useTheme return correct context', () => {
        const { result } = renderHook(() => useTheme());
        expect(result.current).toMatchObject({
            red: '#ff0000',
            green: '#00ff00',
            blue: '#0000ff',
        });
    });
});
