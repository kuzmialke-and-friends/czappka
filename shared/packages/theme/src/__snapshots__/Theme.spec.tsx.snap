// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theme renders correctly 1`] = `
<div>
  test
</div>
`;
